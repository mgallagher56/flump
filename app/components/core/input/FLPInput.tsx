import { useCallback, useId } from 'react';
import type { ChangeEvent, FC } from 'react';

import type { InputProps } from '@chakra-ui/react';
import { Input } from '@chakra-ui/react';
import * as editable from '@zag-js/editable';
import { mergeProps, normalizeProps, useMachine } from '@zag-js/react';

import { css } from 'styled-system/css';

import FLPBox from '../structure/FLPBox';

interface FLPInputProps extends InputProps {
  error?: string;
  label: string;
  isLabelHidden?: boolean;
}

const FLPInput: FC<FLPInputProps> = ({
  colorScheme,
  error,
  flexDirection,
  isLabelHidden,
  label,
  value,
  variant,
  onChange,
  ...props
}) => {
  const [state, send] = useMachine(editable.machine({ id: useId() }));
  const handleOnChange = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      send({ type: 'CHANGE', value: e.target.value });
      onChange && onChange(e);
    },
    [send, onChange]
  );

  const api = editable.connect(state, send, normalizeProps);

  const columnStyles = css({
    display: 'flex',
    flexDirection: 'column',
    gap: props.gap || 2
  });

  const rowStyles = css({
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    gap: props.gap || 4
  });

  const errorStyles = css({
    color: 'red.500',
    fontWeight: 'bold'
  });

  const rootProps = mergeProps({ ...api.rootProps }, { className: flexDirection === 'row' ? rowStyles : columnStyles });
  const inputProps = mergeProps(
    { ...api.inputProps },
    {
      ...props,
      hidden: false,
      variant: variant || 'outline',
      value: value,
      onChange: handleOnChange
    }
  );
  const labelProps = mergeProps({ ...api.labelProps }, { hidden: isLabelHidden });

  return (
    <div {...rootProps}>
      <label {...labelProps}>{label}</label>
      <div {...api.areaProps}>
        <FLPBox display="flex" flexDirection="column" gap={props.gap ?? 2}>
          <Input {...inputProps} defaultValue={undefined} />
          {error && <span className={errorStyles}>{error}</span>}
        </FLPBox>
      </div>
    </div>
  );
};

export default FLPInput;
