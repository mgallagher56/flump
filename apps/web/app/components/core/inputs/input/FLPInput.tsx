import { type FC } from 'react';

import type { InputProps } from '@chakra-ui/react';
import { Input } from '@chakra-ui/react';
import FLPBox from '~/components/core/structure/FLPBox';

import { css } from 'styled-system/css';

interface FLPInputProps extends InputProps {
  error?: string;
  label: string;
  isLabelHidden?: boolean;
}

const FLPInput: FC<FLPInputProps> = ({
  colorScheme,
  error,
  flexDirection,
  isLabelHidden,
  label,
  variant,
  onChange,
  ...props
}) => {
  const columnStyles = css({
    display: 'flex',
    flexDirection: 'column',
    gap: props.gap || 2
  });

  const rowStyles = css({
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    gap: props.gap || 4
  });

  const errorStyles = css({
    color: 'red.500',
    fontWeight: 'bold'
  });

  return (
    <div className={flexDirection === 'row' ? rowStyles : columnStyles}>
      <label htmlFor={label} hidden={isLabelHidden}>
        {label}
      </label>
      <div>
        <FLPBox display="flex" flexDirection="column" gap={props.gap ?? 2}>
          <Input id={label} hidden={false} variant={variant || 'outline'} onChange={onChange} {...props} />
          {error && <span className={errorStyles}>{error}</span>}
        </FLPBox>
      </div>
    </div>
  );
};

export default FLPInput;
