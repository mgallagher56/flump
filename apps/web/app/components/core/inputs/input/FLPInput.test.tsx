import { fireEvent, render } from '@testing-library/react';
import { vi } from 'vitest';

import FLPInput from './FLPInput';

const onChangeMock = vi.fn();

describe('<FLPInput />', () => {
  test('should render as expected', () => {
    const { container } = render(<FLPInput label="Test" onChange={onChangeMock} />);
    expect(container).toMatchSnapshot();
  });
  test('should render as expected with error', () => {
    const { container } = render(<FLPInput label="Test" error="Error" onChange={onChangeMock} />);
    expect(container).toMatchSnapshot();
  });
  test('should render as expected with value', () => {
    const { container } = render(<FLPInput label="Test" value="Value" onChange={onChangeMock} />);
    expect(container).toMatchSnapshot();
  });
  test('should render as expected with value and error', () => {
    const { container } = render(<FLPInput label="Test" value="Value" error="Error" onChange={onChangeMock} />);
    expect(container).toMatchSnapshot();
  });
  test('should render as expected without a label', () => {
    const { container } = render(<FLPInput label="Test" isLabelHidden onChange={onChangeMock} />);
    expect(container).toMatchSnapshot();
  });

  test('should render elements as a row', () => {
    const { container } = render(<FLPInput label="Test" flexDirection="row" onChange={onChangeMock} />);
    expect(container).toMatchSnapshot();
  });

  test('should change the value when typing', () => {
    const { getByLabelText } = render(<FLPInput label="Test" onChange={onChangeMock} />);
    const input = getByLabelText('Test') as HTMLInputElement;
    fireEvent.change(input, { target: { value: 'Test' } });
    expect(onChangeMock).toHaveBeenCalled();
    expect(input.value).toBe('Test');
  });
});
