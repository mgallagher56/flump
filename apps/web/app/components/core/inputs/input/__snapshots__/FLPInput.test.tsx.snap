// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<FLPInput /> > should render as expected 1`] = `
<div>
  <div
    class="d_flex flex_column gap_2"
  >
    <label
      for="Test"
    >
      Test
    </label>
    <div>
      <div
        class="css-3a3hw0"
      >
        <input
          class="chakra-input css-0"
          id="Test"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<FLPInput /> > should render as expected with error 1`] = `
<div>
  <div
    class="d_flex flex_column gap_2"
  >
    <label
      for="Test"
    >
      Test
    </label>
    <div>
      <div
        class="css-3a3hw0"
      >
        <input
          class="chakra-input css-0"
          id="Test"
        />
        <span
          class="text_red.500 fw_bold"
        >
          Error
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<FLPInput /> > should render as expected with value 1`] = `
<div>
  <div
    class="d_flex flex_column gap_2"
  >
    <label
      for="Test"
    >
      Test
    </label>
    <div>
      <div
        class="css-3a3hw0"
      >
        <input
          class="chakra-input css-0"
          id="Test"
          value="Value"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<FLPInput /> > should render as expected with value and error 1`] = `
<div>
  <div
    class="d_flex flex_column gap_2"
  >
    <label
      for="Test"
    >
      Test
    </label>
    <div>
      <div
        class="css-3a3hw0"
      >
        <input
          class="chakra-input css-0"
          id="Test"
          value="Value"
        />
        <span
          class="text_red.500 fw_bold"
        >
          Error
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<FLPInput /> > should render as expected without a label 1`] = `
<div>
  <div
    class="d_flex flex_column gap_2"
  >
    <label
      for="Test"
      hidden=""
    >
      Test
    </label>
    <div>
      <div
        class="css-3a3hw0"
      >
        <input
          class="chakra-input css-0"
          id="Test"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<FLPInput /> > should render elements as a row 1`] = `
<div>
  <div
    class="d_flex flex_row items_center gap_4"
  >
    <label
      for="Test"
    >
      Test
    </label>
    <div>
      <div
        class="css-3a3hw0"
      >
        <input
          class="chakra-input css-0"
          id="Test"
        />
      </div>
    </div>
  </div>
</div>
`;
